---
title: "analysing maf"
author: "Danie Fridljand"
date: "2022"
output: html_document
---


```{r warning=FALSE, message=FALSE}
library(tidyr)
library(IHW)
library(ggplot2)
library(cowplot)
library(scales)
# library(latex2exp)
library(dplyr)
# require(biomaRt)
theme_set(theme_cowplot())

set.seed(1)
options(bitmapType = "cairo")
```

```{r setup, include=FALSE, cache = FALSE}
knitr::opts_knit$set(root_dir = "/g/huber/users/fridljand/R/ihw-forest-paper/") # TODO
knitr::opts_chunk$set(warning = FALSE, message = FALSE)
root_dir <- "/g/huber/users/fridljand/R/ihw-forest-paper/"
```

## function 
```{r}
rejections_na <- function(ihw) { # TODO
  if (is.na(ihw)) {
    return(NA)
  } else {
    rejections(ihw)
  }
}
```

# sampel minor alle frequency

```{r loading_pvalues2}
chr1_df <- readRDS(file.path(root_dir, "data/hqtl_chrom1_chrom2/chr1_subset.Rds")) #%>% dplyr::select(SNP, gene, pvalue)
chr2_df <- readRDS(file.path(root_dir, "data/hqtl_chrom1_chrom2/chr2_subset.Rds")) #%>% dplyr::select(SNP, gene, pvalue)
```

```{r, eval = T}
chr1_maf <- readRDS(file = file.path(root_dir, "data/hqtl_chrom1_chrom2/chr1_maf_sample.rds"))
chr2_maf <- readRDS(file = file.path(root_dir, "data/hqtl_chrom1_chrom2/chr2_maf_sample.rds"))
```

```{r}
chr1_df <- chr1_df %>% left_join(chr1_maf, by = c("SNP" = "refsnp_id"))
chr2_df <- chr2_df %>% left_join(chr2_maf, by = c("SNP" = "refsnp_id"))
```

```{r}
chr1_chr2_df <- rbind(chr1_df %>% mutate(chromosome = 1), 
                      chr2_df %>% mutate(chromosome = 2))
```

```{r}
rm(chr1_df, chr2_df, chr1_maf, chr2_maf)
```

```{r}
pval_threshold <- 10^(-10)
chr1_chr2_df <-chr1_chr2_df %>% filter(pvalue <= pval_threshold)
```

Hence, the censored p-values follow the model 
\begin{eqnarray}
&P_i \mid H_i = 0 \sim U[0,10^(-4)]
\end{eqnarray}

However, all multiple testing used in this vigniette assume 
\begin{eqnarray}
&P_i \mid H_i = 0 \sim U[0,1]
\end{eqnarray}
We address this gap by artificially inflating the p-values:  
```{r}
chr1_chr2_df <- chr1_chr2_df %>% mutate(pvalue = pvalue / pval_threshold)
```

## compare minor alle frequency and standard deviation of effect size estimate

```{r}
chr1_chr2_df <- chr1_chr2_df %>% mutate(beta_se = beta/tstat)
ggplot(chr1_chr2_df,
       aes(x = minor_allele_freq, y = beta_se)) +
  scale_y_log10()+
  geom_hex() + #bins = 60
  #geom_point()+
  facet_wrap(~chromosome, ncol = 2)
```

## analyse distribution of p-values 

```{r, eval = FALSE, include = FALSE}
labs <- cut_number(chr1_chr2_df$minor_allele_freq, 8)
upper = as.numeric( sub("[^,]*,([^]]*)\\]", "\\1", labs) )
cutoffs = unique(upper)
```

```{r, eval = FALSE, include = FALSE}
cutoffs <- c(0.06, 0.07, 0.08, 0.09, 0.1, 0.3, 0.4)
```

```{r}
chr1_chr2_df$mafGroup <- groups_by_filter(chr1_chr2_df$minor_allele_freq, 8)
cutoffs <- chr1_chr2_df %>%
  group_by(mafGroup) %>%
  summarise(cutoffs = max(minor_allele_freq)) %>%
  select(cutoffs) %>%
  unlist
```

`chr1_chr2_df` can be found under https://stanfordmedicine.box.com/s/86ng19kdrueak3wncjf1ubphmqacrkrp. 

```{r}
# set up data frame for ggplotting
ggplot(chr1_chr2_df,
       aes(x = minor_allele_freq, y = -log10(pvalue))) +
  geom_hex(bins = 60) +
  geom_vline(xintercept=cutoffs, linetype="dashed") +
  # geom_point()+
  ylab(expression(-log[10] ~ p)) +
  facet_wrap(~chromosome, ncol = 2)
```

```{r}
cutoffs_extend <- c(0, cutoffs)

#chr1_chr2_df$mafGroup <- cut(chr1_chr2_df$minor_allele_freq, cutoffs_extend)
table(chr1_chr2_df$mafGroup, chr1_chr2_df$chromosome)
```
Enough observations in each bin

## pure minor_allele_freq
```{r}
ggplot(chr1_chr2_df %>% filter(chromosome == 1),
       aes(x = pvalue)) +
      geom_histogram(boundary = 0, bins = 60)
```

```{r}
ggplot(chr1_chr2_df %>% filter(chromosome == 1), 
       aes(x = pvalue)) +
  geom_histogram(boundary = 0) +
  facet_wrap(~mafGroup, nrow = 2)
```
## run IHW 
without regularisation
```{r}
#alphas <- seq(0.001, 0.01, length.out = 5)
alphas <-c(c(1,3,5,7) %o% 10^(-2:-5))
alphas <-c(c(1,7) %o% 10^(-2:-4))
#alphas <- 0.05
```

```{r}
ihw_quantiles <- lapply(alphas, function(alpha_i) {
  ihw_quantile_i <- ihw(
    pvalues = chr1_chr2_df$pvalue,
    #covariates = chr1_chr2_df$mafGroup,
    covariates = chr1_chr2_df$minor_allele_freq,
    alpha = alpha_i,
    folds = chr1_chr2_df$chromosome#,
    #lambdas = Inf
  )
})
```

```{r}
ihw_quantiles_no_regul <- lapply(alphas, function(alpha_i) {
  ihw_quantile_i <- ihw(
    pvalues = chr1_chr2_df$pvalue,
    #covariates = chr1_chr2_df$mafGroup,
    covariates = chr1_chr2_df$minor_allele_freq,
    alpha = alpha_i,
    folds = chr1_chr2_df$chromosome,
    lambdas = Inf
  )
})
```

```{r}
bh_rejections <- sapply(alphas, function(alpha_i) {
  sum(p.adjust(na.exclude(chr1_chr2_df$pvalue), method = "BH") <= alpha_i)
})
```

```{r}
ihw_quantile_rejections <- sapply(ihw_quantiles, rejections_na)
ihw_quantiles_no_regul_rejections <- sapply(ihw_quantiles_no_regul, rejections_na)


result_bind <- data.frame(alpha = alphas,
           ihw_quantile_rejections = ihw_quantile_rejections,
           bh_rejections = bh_rejections,
           ihw_quantiles_no_regul_rejections = ihw_quantiles_no_regul_rejections)

result_bind <- result_bind %>%
  pivot_longer(c("ihw_quantile_rejections","bh_rejections", "ihw_quantiles_no_regul_rejections"),
               names_to = "method",
               values_to = "rejections") 
```

## plot IHW result
```{r}
hqtl_rejections <- ggplot(result_bind, aes(x = alpha, y = rejections, col = method)) +
  geom_line() +
  geom_point(size = 2) +
  ylab("Number of rejections") +
  xlab("Target FDR") +
  theme(legend.position="bottom")+
 guides(
  color = guide_legend(title = "Method", legend.text = element_text(size = 4),  nrow = 2),
  shape = "none"
 ) +
# scale_x_continuous(breaks = seq(0, 1000, by = 250)) +
  scale_x_continuous(breaks = unique(result_bind$alpha), trans = "log") 
  
hqtl_rejections
```
```{r}
hqtl_rejections + scale_y_continuous(limits = c(0,NA))
```

## dive into weights
```{r}
ihw_quantiles_first <- ihw_quantiles[[1]]
range(IHW::thresholds(ihw_quantiles_first))
```

```{r}
ihw_quantiles_first <- ihw_quantiles[[1]]
weights(ihw_quantiles_first, levels_only = TRUE)
```
```{r}
plot(ihw_quantiles_first)
```
```{r}
plot(ihw_quantiles_no_regul[[1]])
```

How can I dive deeper in the last plot and understand it?

```{r}
saveRDS(chr1_chr2_df, "chr1_chr2_df.Rds")
```

