---
title: "Read in BMI GIANT GWAS meta-analysis data"
author: "Simina Boca, Jeff Leek"
date: "`r Sys.Date()`"
output: rmarkdown::pdf_document
vignette: >
  %\VignetteIndexEntry{swfdr}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

<!--
%% \VignetteEngine{knitr::rmarkdown}
-->

```{r global_options, include=FALSE}
knitr::opts_chunk$set(fig.path='Figures/')
```

```{r style, echo = FALSE, results = 'asis'}
BiocStyle::markdown()
```

Load the relevant libraries first:

```{r,message=FALSE}
library(readr)
library(dplyr)
library(reshape2)
library(curl)
library(Hmisc)
library(here)
library(biomaRt)
```

Now load the data (first need to download the file from https://www.broadinstitute.org/collaboration/giant/index.php/GIANT_consortium_data_files#GWAS_Anthropometric_2015_BMI):

https://portals.broadinstitute.org/collaboration/giant/images/f/f0/All_ancestries_SNP_gwas_mc_merge_nogc.tbl.uniq.gz
```{r,cache=TRUE}
##change the directory name
tot = read_tsv(here("boca_leek/All_ancestries_SNP_gwas_mc_merge_nogc.tbl.uniq"))
head(tot)
dim(tot)
```


Do some basic data filtering (to remove missing values) and data cleaning (switch to minor allele from effect allele, add variable with frequency intervals):
```{r,cache=TRUE}
tot = filter(tot,!is.na(Freq1.Hapmap))
dim(tot)
##get distribution of sample size for the SNP associations
quantile(tot$N)
##change the frequency to the minor allele frequency
changeWhich <- tot$Freq1.Hapmap > 0.5
tot$Freq1.Hapmap[changeWhich] <- 1-tot$Freq1.Hapmap[changeWhich]
colnames(tot)[colnames(tot)=="Freq1.Hapmap"] <- "Freq_MAF_Hapmap"
##also need to change the coefficient sign for these alleles
tot$b[changeWhich] <- -tot$b[changeWhich]
tot = mutate(tot,Freq_MAF_Int_Hapmap=cut2(Freq_MAF_Hapmap, g=3)) ##added a new variable with frequency intervals for HapMap CEU frequencies
table(tot$Freq_MAF_Int_Hapmap)
##also get the distributions of MAFs
quantile(tot$Freq_MAF_Hapmap)
```

## add annotation rtracklayer

```{r}

```


## Additional annotation biomaRt
```{r, eval = FALSE}
ensembl <- biomaRt::useMart("ENSEMBL_MART_SNP", dataset = "hsapiens_snp")
#download data
annotations_locke <- biomaRt::getBM(
  attributes = c("refsnp_id", "minor_allele_freq", "chr_name", "chrom_start","chrom_end"),
  filters = "snp_filter", values = unique(tot$SNP),
  mart = ensembl, uniqueRows = TRUE
)
saveRDS(annotations_locke, "annotations_locke.Rds")
```

```{r}
annotations_locke <- readRDS(here("boca_leek/annotations_locke.Rds"))
head(annotations_locke)
```

```{r}
annotations_locke <- annotations_locke %>%
  group_by(refsnp_id) %>%
  slice(1) %>%
  ungroup()
```

```{r}
tot_annot <- inner_join(annotations_locke, tot,  by = c("refsnp_id"="SNP"))
tot_annot <-  tot_annot %>%
  mutate(minor_allele_freq = as.double(minor_allele_freq),
         chrom_start = as.integer(chrom_start),
         chrom_end = as.integer(chrom_end))
head(tot_annot)
```

```{r}
tot_annot %>%
  ggplot(aes(x = Freq_MAF_Hapmap, y = minor_allele_freq)) +
  geom_hex(bins = 100)
```

```{r}
tot_annot <- tot_annot %>%
  filter(!is.na(minor_allele_freq))
```

```{r}
# Step 2: Calculate the quantiles
quantiles_freq <- quantile(tot_annot$Freq_MAF_Hapmap, probs = seq(0, 1, by = 0.01))
quantiles_maf <- quantile(tot_annot$minor_allele_freq, probs = seq(0, 1, by = 0.01))

# Step 3: Create a data frame with the quantiles
quantiles_df <- data.frame(Freq_MAF_Hapmap = quantiles_freq, minor_allele_freq = quantiles_maf)

# Step 4: Create a new plot
ggplot(quantiles_df, aes(x = Freq_MAF_Hapmap, y = minor_allele_freq)) +
  geom_point() +
  geom_abline(slope=1, intercept = 0)
```

Save in a data frame in a .RData file:

```{r}
BMI_GIANT_GWAS <- tot_annot
save(x=BMI_GIANT_GWAS, file="BMI_GIANT_GWAS.RData")
```

### Session Information

```{r}
devtools::session_info()
```