---
title: "Read in BMI GIANT GWAS meta-analysis data"
author: "Simina Boca, Jeff Leek"
date: "`r Sys.Date()`"
output: rmarkdown::pdf_document
vignette: >
  %\VignetteIndexEntry{swfdr}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r global_options, include=FALSE}
knitr::opts_chunk$set(fig.path='Figures/')
```

```{r style, echo = FALSE, results = 'asis', message = FALSE}
BiocStyle::markdown()
```

Load the relevant libraries first:

```{r,message=FALSE}
library(readr)
library(dplyr)
library(reshape2)
library(curl)
library(Hmisc)
library(here)
library(biomaRt)
library(rtracklayer)
library(tidygenomics)
library(purrr)
library(tidyr)
```

Now load the data (first need to download the file from <https://www.broadinstitute.org/collaboration/giant/index.php/GIANT_consortium_data_files#GWAS_Anthropometric_2015_BMI>):

<https://portals.broadinstitute.org/collaboration/giant/images/f/f0/All_ancestries_SNP_gwas_mc_merge_nogc.tbl.uniq.gz>

```{r,cache=TRUE}
##change the directory name
tot = read_tsv(here("boca_leek/All_ancestries_SNP_gwas_mc_merge_nogc.tbl.uniq"))
head(tot)
dim(tot)
```

Do some basic data filtering (to remove missing values) and data cleaning (switch to minor allele from effect allele, add variable with frequency intervals):

```{r,cache=TRUE}
tot = filter(tot,!is.na(Freq1.Hapmap))
dim(tot)
##get distribution of sample size for the SNP associations
quantile(tot$N)
##change the frequency to the minor allele frequency
changeWhich <- tot$Freq1.Hapmap > 0.5
tot$Freq1.Hapmap[changeWhich] <- 1-tot$Freq1.Hapmap[changeWhich]
colnames(tot)[colnames(tot)=="Freq1.Hapmap"] <- "Freq_MAF_Hapmap"
##also need to change the coefficient sign for these alleles
tot$b[changeWhich] <- -tot$b[changeWhich]
tot = mutate(tot,Freq_MAF_Int_Hapmap=cut2(Freq_MAF_Hapmap, g=3)) ##added a new variable with frequency intervals for HapMap CEU frequencies
table(tot$Freq_MAF_Int_Hapmap)
##also get the distributions of MAFs
quantile(tot$Freq_MAF_Hapmap)
```

## Additional annotation biomaRt

```{r, eval = FALSE}
ensembl <- biomaRt::useMart("ENSEMBL_MART_SNP", dataset = "hsapiens_snp")
#download data
annotations_locke <- biomaRt::getBM(
  attributes = c("refsnp_id", "minor_allele_freq", "chr_name", "chrom_start","chrom_end"),
  filters = "snp_filter", values = unique(tot$SNP),
  mart = ensembl, uniqueRows = TRUE
)
saveRDS(annotations_locke, "annotations_locke.Rds")
```

```{r}
annotations_locke <- readRDS(here("boca_leek/annotations_locke.Rds"))
head(annotations_locke)
```

```{r}
annotations_locke <- annotations_locke %>%
  group_by(refsnp_id) %>%
  slice(1) %>%
  ungroup()
```

```{r}
tot_annot <- inner_join(annotations_locke, tot,  by = c("refsnp_id"="SNP"))
tot_annot <-  tot_annot %>%
  mutate(minor_allele_freq = as.double(minor_allele_freq),
         chrom_start = as.integer(chrom_start),
         chrom_end = as.integer(chrom_end))
head(tot_annot)
```

```{r}
tot_annot %>%
  ggplot(aes(x = Freq_MAF_Hapmap, y = minor_allele_freq)) +
  geom_hex(bins = 100)
```

```{r}
tot_annot <- tot_annot %>%
  filter(!is.na(minor_allele_freq))
```

```{r}
# Step 2: Calculate the quantiles
quantiles_freq <- quantile(tot_annot$Freq_MAF_Hapmap, probs = seq(0, 1, by = 0.01))
quantiles_maf <- quantile(tot_annot$minor_allele_freq, probs = seq(0, 1, by = 0.01))

# Step 3: Create a data frame with the quantiles
quantiles_df <- data.frame(Freq_MAF_Hapmap = quantiles_freq, minor_allele_freq = quantiles_maf)

# Step 4: Create a new plot
ggplot(quantiles_df, aes(x = Freq_MAF_Hapmap, y = minor_allele_freq)) +
  geom_point() +
  geom_abline(slope=1, intercept = 0)
```

```{r}
BMI_GIANT_GWAS <- tot_annot
save(x=BMI_GIANT_GWAS, file="BMI_GIANT_GWAS.RData")
```

## add annotation rtracklayer

```{r}
annotations_locke <- readRDS(here("boca_leek/annotations_locke.Rds"))
annotations_locke <- annotations_locke %>%
  mutate(chrom_start = as.numeric(chrom_start),
         chrom_end = as.numeric(chrom_end))
head(annotations_locke)
```

```{r}
annotations_locke_filtered <- annotations_locke %>%
  filter(chr_name %in% c(1,10)) #c(1,10,11,12,13)
```

```{r}
ranges_by_chr <- annotations_locke_filtered %>%
  group_by(chr_name) %>%
  summarise(min(chrom_start),
            max(chrom_end))
ranges_by_chr
```

```{r}
chr1_grange <- GRanges("chr1", IRanges(100000723 *0.8, 99997866 *1.2))
chr10_grange <- GRanges("chr10", IRanges(100002012 *0.8, 99999225 *1.2))
```

```{r}
mySession <- browserSession()
# can also use hg38 the newer human genome
genome(mySession) <- "hg19"
track.names.all <- trackNames(ucscTableQuery(mySession))
```

```{r}
track_table_ucsc <- tribble(
  ~track, ~table,
  "wgEncodeRegMarkH3k27ac", "wgEncodeBroadHistoneGm12878H3k27acStdSig",
  "wgEncodeRegMarkH3k4me3", "wgEncodeBroadHistoneGm12878H3k4me3StdSig",
  "wgEncodeUwTfbs", "wgEncodeUwTfbsGm12878CtcfStdHotspotsRep1",
  "wgEncodeUwDnase", "wgEncodeUwDnaseGm12878HotspotsRep1",
  "wgEncodeRegMarkH3k4me1", "wgEncodeBroadHistoneGm12878H3k4me1StdSig"
)
```

```{r}
#for each additional dimension, 
ucsc_additional_covariates <- pmap(track_table_ucsc, 
                function(track, table){
                  chr1_additional_covariates <- ucscTableQuery(mySession, 
                                                               track, 
                                                               table, 
                                                               range = chr1_grange) %>%
                    getTable() %>%
                    #add table and track information
                    mutate(chr_name = "1")
                  
                  chr10_additional_covariates <- ucscTableQuery(mySession, 
                                                               track, 
                                                               table, 
                                                               range = chr10_grange) %>%
                    getTable() %>%
                    
                    mutate(chr_name = "10")
                  
                  #combine both chromosomes
                  ucsc_additional_covariates <- rbind(chr1_additional_covariates, chr10_additional_covariates) 
                  
                  if("chromStart" %in% names(ucsc_additional_covariates)){
                    ucsc_additional_covariates <- ucsc_additional_covariates %>%
                      dplyr::select(chr_name, start = chromStart, end = chromEnd, value = signalValue)
                  } 
                  #add table and track information
                  ucsc_additional_covariates %>%
                        mutate(track = track, table = table)
                })

```

join via genomic region

```{r}
ucsc_additional_covariates <- map_dfr(ucsc_additional_covariates, 
                function(ucsc_additional_covariates_i){
                  annotations_locke_filtered %>%
                    genome_left_join_closest(ucsc_additional_covariates_i, by = c("chr_name","chrom_start" = "start","chrom_end" = "end"))
                })

```

```{r}
ucsc_additional_covariates <- ucsc_additional_covariates %>%
  select(refsnp_id, chr_name= chr_name.x,value, track, table) %>%
  tidyr::unite("track_table", track:table) 

head(ucsc_additional_covariates)
```

```{r}
ucsc_additional_covariates <- ucsc_additional_covariates %>%
  mutate(covariate_abbr = forcats::fct_recode(track_table,
  "covariate1"="wgEncodeRegMarkH3k27ac_wgEncodeBroadHistoneGm12878H3k27acStdSig",
  "covariate2"= "wgEncodeRegMarkH3k4me3_wgEncodeBroadHistoneGm12878H3k4me3StdSig",
  "covariate3"= "wgEncodeUwTfbs_wgEncodeUwTfbsGm12878CtcfStdHotspotsRep1",
  "covariate4" = "wgEncodeUwDnase_wgEncodeUwDnaseGm12878HotspotsRep1",
  "covariate5" = "wgEncodeRegMarkH3k4me1_wgEncodeBroadHistoneGm12878H3k4me1StdSig"
  ))

ucsc_additional_covariates <- ucsc_additional_covariates %>%
  select(-c(track_table)) %>%
  pivot_wider(
    names_from = covariate_abbr,
    values_from = value,
    values_fn = mean
  )

head(ucsc_additional_covariates)
```

Save in a data frame in a .RData file:

```{r}
save(x = ucsc_additional_covariates, file = here("boca_leek/data/locke_ucsc_additional_covariates.RData"))
```

### Session Information

```{r}
devtools::session_info()
```
